daemon/internal/newrelic/app.go

```diff
@@ -338,3 +340,75 @@ func (app *App) Inactive(threshold time.Duration) bool {
 	}
 	return time.Since(app.LastActivity) > threshold
 }
+
+// filter seen php packages data to avoid sending duplicates
+//
+// the `App` structure contains a map of PHP Packages the reporting
+// application has encountered.
+//
+// the map of packages should persist for the duration of the
+// current connection
+//
+// takes the `PhpPackages.data` byte array as input and unmarshals
+// into an anonymous interface array
+//
+// the JSON format received from the agent is:
+// 		[["package_name","version",{}],...]
+//
+// for each entry, assign the package name and version to the `PhpPackagesKey`
+// struct and use the key to verify data does not exist in the map. If the
+// key does not exist, add it to the map and the array of 'new' packages.
+//
+// convert the array of 'new' packages into a byte array representing
+// the expected data that should match input, minus the duplicates.
+func (app *App) filterPhpPackages(data []byte) []byte {
+	if data == nil {
+		return nil
+	}
+
+	var pkgKey PhpPackagesKey
+	var newPkgs []PhpPackagesKey
+	var x []interface{}
+
+	err := json.Unmarshal(data, &x)
+	if nil != err {
+		log.Errorf("failed to unmarshal php package json: %s", err)
+		return nil
+	}
+
+	for _, pkgJson := range x {
+		pkg, _ := pkgJson.([]interface{})
+		if len(pkg) != 3 {
+			log.Errorf("invalid php package json structure: %+v", pkg)
+			return nil
+		}
+		pkgKey = PhpPackagesKey{pkg[0].(string), pkg[1].(string)}
+		_, ok := app.PhpPackages[pkgKey]
+		if !ok {
+			app.PhpPackages[pkgKey] = struct{}{}
+			newPkgs = append(newPkgs, pkgKey)
```

Comment:
Is the newPkgs needed here?  Couldn't we just directly be writing to the `buf` variable instead?

Reply:
It's possible, but in my opinion not very clean. 
The buffer that outputs the result requires a very specific format that is easier to maintain if it's separated into it's own logic rather than injected into this for loop that's trying to parse out the data from the `byte` array.
Additionally, the logic to populate the buffer checks to verify that `newPkgs` is not `nil`, which is not possible using just the `buf` variable without some less than elegant logic.

Reply:
The tradeoff to consider is the additional overhead.

Couldn't you just check `buf` len to determine whether to return nil?

Reply:
In terms of performance, since this should only be happening once per harvest and only contain a handful of elements (the frameworks x versions we instrument) I don't anticipate this having any impact to overhead.
I prefer to keep the responsibility of things as atomic as possible. Currently this for loop is responsible for consuming the `byte` array and sorting out the "new" vs "old" packages. Constructing the resulting `byte` array is a separate task altogether, so I prefer to keep it decoupled from the parsing/sorting logic. This has the benefit of:
1. code readability- one block of code is responsible for one "job" rather than all jobs
2. ease of debugging- you can easily verify the result of the loop via checking the `newPkgs` variable, and the result of re-marshalling by checking the output of `buf`
3. ease of adding/changing functionality in the future- decoupling data parsing and data formatting isolates responsibilities makes modifying certain behavior or adding new behavior less invasive and more extensible in the long-term.

Reply:
It would probably be cleaner/more efficient if the agent sent package/version pairs instead of a JSON string.  But I agree that the number of packages per request is bounded by a small number and I suspect the vast majority of the time there will be no new packages so `newPkgs` will be `nil` most the time and the code below won't execute.  

Could be interesting to wrap the loop starting on line 379 with a timer to see how long it takes with different size JSON payloads just to put to rest any concern this could be time consuming with realistic payloads.

Reply:
The more I think about it I like the idea of making `PhpPackagesKey` a `struct` in the `protocol.fbs` and just passing it over from the agent that way which can be trivially decoded.

Reply:
This is outside the scope of this PR.